invoice/src/main/java/org/killbill/billing/invoice/dao/DefaultInvoiceDao.java

```diff
@@ -963,8 +959,19 @@ public boolean apply(final InvoicePaymentModelDao input) {
                     if (existingAttempt == null) {
                         createAndRefresh(transactional, invoicePayment, context);
                     } else {
+                        final UUID updatedPaymentId;
+                        if (invoicePayment.getPaymentId() == null) {
+                            // Likely invalid state (https://github.com/killbill/killbill/issues/1230) but go ahead nonetheless to run the payment state machine validations
+                            updatedPaymentId = existingAttempt.getPaymentId();
+                        } else if (existingAttempt.getPaymentId() == null) {
+                            updatedPaymentId = invoicePayment.getPaymentId();
+                        } else {
+                            Preconditions.checkState(invoicePayment.getPaymentId().equals(existingAttempt.getPaymentId()),
+                                                     "PaymentAttempt cannot change paymentId: attemptId=%s, newInvoicePaymentId=%s, existingPaymentId=%s", existingAttempt.getId(), invoicePayment.getPaymentId(), existingAttempt.getPaymentId());
+                            updatedPaymentId = invoicePayment.getPaymentId();
+                        }
```

Comment:
Wouldn't it be simpler to do:
```
if (invoicePayment.getPaymentId() != null) {
   Preconditions.checkState(....);
    updatedPaymentId = invoicePayment.getPaymentId();
} else { 
  // default to paymentAttempt
  updatedPaymentId = existingAttempt.getPaymentId();
}
```

Reply:
Not quite - `Preconditions` is only valid if `existingAttempt.getPaymentId() != null`.

But regardless, I wanted to call out the invalid branch explicitly (source of the bug).